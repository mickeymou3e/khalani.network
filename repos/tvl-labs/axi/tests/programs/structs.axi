struct Foo {
    x: Nat,
    y: Nat,
    z: Nat,
}

!claim(forall x1: Nat, y1: Nat, z1: Nat, x2: Nat, y2: Nat, z2: Nat. (
    Foo { x: x1, y: y1, z: z1 } = Foo { x: x2, y: y2, z: z2 } ==>
    x1 = x2 & y1 = y2 & z1 = z2
));

struct Bar {
    f: Foo,
}

let foo := Foo { x: 3, z: 5, y: 4 };

# Test that the order of fields doesn't matter.
!eq_chain(foo, Foo { z: 5, x: 3, y: 4 });

foo.x;

let bar := Bar { f: foo };

bar.f.x;

# Test elimination of Foo {}.field
!eq_chain(5, foo.z);
!eq_chain(bar.f.y, foo.y, 4);

!prove_by_eval(bar.f.y = 4);
