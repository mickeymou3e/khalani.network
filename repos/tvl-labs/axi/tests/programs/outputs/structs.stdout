Result: Nothing
Result: Proved(forall Nat. forall Nat. forall Nat. forall Nat. forall Nat. forall Nat. if(=[Foo.mk[@5:Nat, @4:Nat, @3:Nat]):Foo, Foo.mk[@2:Nat, @1:Nat, @0:Nat]):Foo], and(and(=[@5:Nat, @2:Nat], =[@4:Nat, @1:Nat]), =[@3:Nat, @0:Nat])))
Result: Nothing
Result: Nothing
Result: Proved(=[Foo.mk[3, 4, 5]):Foo, Foo.mk[3, 4, 5]):Foo])
Result: Term(Foo.x[Foo.mk[3, 4, 5]):Foo]):Nat)
Result: Nothing
Result: Term(Foo.x[Bar.f[Bar.mk[Foo.mk[3, 4, 5]):Foo]):Bar]):Foo]):Nat)
Result: Proved(=[5, Foo.z[Foo.mk[3, 4, 5]):Foo]):Nat])
Result: Proved(=[Foo.y[Bar.f[Bar.mk[Foo.mk[3, 4, 5]):Foo]):Bar]):Foo]):Nat, 4])
Result: Proved(=[Foo.y[Bar.f[Bar.mk[Foo.mk[3, 4, 5]):Foo]):Bar]):Foo]):Nat, 4])
